rust/tw_keypair/src/ed25519/modifications/cardano/extended_private.rs

```diff
@@ -0,0 +1,152 @@
+        &self,
+        public: &ExtendedPublicKey<H>,
+        message: &[u8],
+    ) -> Result<Signature, Error> {
+        self.key
+            .expanded_key
+            .sign_with_pubkey(public.key_for_signing(), message)
+    }
+}
+
+impl<H: Hasher512> SigningKeyTrait for ExtendedPrivateKey<H> {
+    type SigningMessage = Vec<u8>;
+    type Signature = Signature;
+
+    fn sign(&self, message: Self::SigningMessage) -> Result<Self::Signature, Error> {
+        self.sign_with_public_key(&self.public(), message.as_slice())
+    }
+}
+
+impl<H: Hasher512> ToBytesZeroizing for ExtendedPrivateKey<H> {
+    fn to_zeroizing_vec(&self) -> Zeroizing<Vec<u8>> {
+        let mut res = self.key.to_zeroizing_vec();
+        res.extend_from_slice(self.second_key.to_zeroizing_vec().as_slice());
+        res
+    }
+}
+
+impl<'a, H: Hasher512> TryFrom<&'a [u8]> for ExtendedPrivateKey<H> {
+    type Error = Error;
+
+    fn try_from(bytes: &'a [u8]) -> Result<Self, Self::Error> {
+        if bytes.len() != Self::LEN {
+            return Err(Error::InvalidSecretKey);
+        }
+        let key = ExtendedSecretPart::try_from(&bytes[Self::KEY_RANGE])?;
+        let second_key = ExtendedSecretPart::try_from(&bytes[Self::SECOND_KEY_RANGE])?;
+
+        Ok(ExtendedPrivateKey { key, second_key })
+    }
+}
+
+/// Implement `str` -> `ExtendedPrivateKey<H>` conversion for test purposes.
+impl<H: Hasher512> From<&'static str> for ExtendedPrivateKey<H> {
+    fn from(hex: &'static str) -> Self {
+        // There is no need to zeroize the `data` as it has a static lifetime (so most likely included in the binary).
+        let data = hex::decode(hex).expect("Expected a valid Secret Key hex");
+        ExtendedPrivateKey::try_from(data.as_slice()).expect("Expected a valid Secret Key")
+    }
+}
+
+#[derive(ZeroizeOnDrop)]
+struct ExtendedSecretPart<H: Hasher512> {
+    secret: H256,
+    extension: H256,
+    chain_code: H256,
+    /// An expanded secret key obtained from [`ExtendedSecretPart::secret`]
+    /// and [`ExtendedSecretPart::extension`].
+    /// It's used to generate a public key and sign messages.
+    expanded_key: ExpandedSecretKey<H>,
+}
+
+/// cbindgen:ignore
+impl<H: Hasher512> ExtendedSecretPart<H> {
+    const LEN: usize = 96;
+    const SECRET_RANGE: Range<usize> = 0..32;
+    const EXTENSION_RANGE: Range<usize> = 32..64;
+    const CHAIN_CODE_RANGE: Range<usize> = 64..96;
+}
+
+impl<H: Hasher512> ToBytesZeroizing for ExtendedSecretPart<H> {
+    fn to_zeroizing_vec(&self) -> Zeroizing<Vec<u8>> {
+        let mut res = Vec::with_capacity(H256::len() * 3);
+        res.extend_from_slice(self.secret.as_slice());
+        res.extend_from_slice(self.extension.as_slice());
+        res.extend_from_slice(self.chain_code.as_slice());
+        Zeroizing::new(res)
+    }
+}
+
+impl<'a, H: Hasher512> TryFrom<&'a [u8]> for ExtendedSecretPart<H> {
+    type Error = Error;
+
+    fn try_from(bytes: &'a [u8]) -> Result<Self, Self::Error> {
```

Comment:
What format is this? Ie. `<SECRET><EXTENSION><CHAIN_CODE>`, is this Cardano specific?

Reply:
I reproduced exactly the same scheme as we have in C++:
https://github.com/trustwallet/wallet-core/blob/b122c5c8706edc07cf8c6bac4ce2119ab666da18/src/PrivateKey.cpp#L96-L109